src/util/binparse/fd_bin_parse.c

```diff
@@ -0,0 +1,249 @@
+#include "fd_bin_parse.h"
+
+/* Whereas the fd_slice*() APIs ensure that the bounds of the src input slice are not violated,
+   these macros are responsible for ensuring that we do not read beyond the bounds of the particular
+   input message blob that we're parsing. */
+#define INPUT_BLOB_SZ_REMAINING (ctx->input_blob_sz - ((ulong)(ctx->src.cur - ctx->pre_parse_src_cur)))
+
+#define CHECK_INPUT_BLOB_SZ_REMAINING(sz)  do {                               \
+  if( FD_UNLIKELY( INPUT_BLOB_SZ_REMAINING<sz ) ) {                           \
+    FD_LOG_WARNING(( "input blob too short" ));                               \
+    return 0;                                                                 \
+  }                                                                           \
+} while(0)
+
+void fd_bin_parse_init( fd_bin_parse_ctx_t * ctx,
+                            void           * src,
+                            ulong            src_sz,
+                            void           * dst,
+                            ulong            dst_sz ) {
+  ctx->src.cur = src;
+  ctx->src.end = (uchar *)src + src_sz;
+  ctx->dst.cur = dst;
+  ctx->dst.end = (uchar *)dst + dst_sz;
+  ctx->pre_parse_src_cur = ctx->src.cur;
+  ctx->pre_parse_dst_cur = ctx->dst.cur;
+  ctx->input_blob_sz = 0;
+}
+
+void
+fd_bin_parse_set_input_blob_size( fd_bin_parse_ctx_t * ctx,
+                                   ulong                input_blob_sz ) {
+  ctx->input_blob_sz = input_blob_sz;
+  ctx->pre_parse_src_cur = ctx->src.cur;
+  ctx->pre_parse_dst_cur = ctx->dst.cur;
+}
+
+int
+fd_bin_parse_was_entire_input_blob_consumed( fd_bin_parse_ctx_t * ctx ) {
+  return ( (ulong)(ctx->src.cur-ctx->pre_parse_src_cur) )==ctx->input_blob_sz;
+}
+
+/* If parsing of a gossip message succeeds, we advance the dst `cur` up to beyond
+   where the parsed structures were written to, and there's no need to update the src
+   `cur` pointer because this occurred fully during parsing. This invariant is further
+   checked by a call to `fd_bin_parse_was_entire_input_blob_consumed` */
+void
+fd_bin_parse_update_state_succeeded( fd_bin_parse_ctx_t * ctx,
+                                     ulong                data_written_sz ) {
+  fd_slice_increment_slice( &(ctx->dst), data_written_sz );
+}
+
+/* If parsing of a gossip message fails, we advance the src `cur` beyond the
+   bad payload so as to move onto the next one, and we rewind the dst `cur` back to where
+   it was before the parse began. */
+void
+fd_bin_parse_update_state_failed( fd_bin_parse_ctx_t * ctx ) {
+  ctx->src.cur = ctx->pre_parse_src_cur + ctx->input_blob_sz;
+  ctx->dst.cur = ctx->pre_parse_dst_cur;
+}
+
+int
+fd_bin_parse_is_enough_space_in_src( fd_bin_parse_ctx_t * ctx,
+                                     ulong                sz   ) {
+  return ( FD_LIKELY( ctx->input_blob_sz>=sz && fd_slice_is_enough_space( &(ctx->src), sz ) ) );
```

Comment:
src/util/binparse/fd_bin_parse.c:64:12: error: implicit conversion loses integer precision: 'long' to 'int' [-Werror,-Wshorten-64-to-32]
  return ( FD_LIKELY( ctx->input_blob_sz>=sz && fd_slice_is_enough_space( &(ctx->src), sz ) ) );
  ~~~~~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/util/binparse/../math/../bits/../fd_util_base.h:473:24: note: expanded from macro 'FD_LIKELY'
#define FD_LIKELY(c)   __builtin_expect( !!(c), 1L )
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.